### PR TITLE
[DP-133] asciidump - BGP Stats

### PR DESCRIPTION
**Pull request by tacocat21**
Related pipedream changes: https://github.com/deepfield/pipedream/pull/10414

### CODE REVIEW

File: nest/proto.c [hunk 1]
```diff
@@ -89,12 +97,58 @@ proto_log_state_change(struct proto *p)
...
+ */
+static void
+notify_bgp_stat_daemon(struct proto *p){
+  if (!p)
+  {
+    return;
+  }
+  char * state = proto_state_name(p);
+  char * name = p->name;
+  char * message = (char*)malloc(strlen(state) + strlen(name)+ 20);
```
Comment (maxmzkr) [inline nest/proto.c]: I'm not sure what sandy would think is best, but we might want the packet to be structured a little big more

we might want to have it start with a version number, message type followed by state length and followed by name length.

Even if we don't want to do a more traditional type of udp packet, then we should still ensure a message type is encoded some where

Comment (maxmzkr) [inline nest/proto.c]: make sure you free this memory

File: nest/proto.c [hunk 2]
```diff
@@ -89,12 +97,58 @@ proto_log_state_change(struct proto *p)
...
+send_message_to_bgp_daemon(char* message){
+    int sockfd;
+    struct sockaddr_in ai_addr;
+    int numbytes;
+    ai_addr.sin_family = AF_INET;
+    ai_addr.sin_port = htons(BGP_PORT);
+    ai_addr.sin_addr.s_addr = inet_addr(BGP_ADDR);
+
+    if ((sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) == -1) {
+        return -1;
```
Comment (maxmzkr) [inline nest/proto.c]: might be good to log the error message here

File: nest/proto.c [hunk 3]
```diff
@@ -89,12 +97,58 @@ proto_log_state_change(struct proto *p)
...
+    }
+
+    if ((numbytes = sendto(sockfd, message, strlen(message), 0, (struct sockaddr *) &ai_addr, sizeof(ai_addr))) == -1) {
+        return -1;
```
Comment (maxmzkr) [inline nest/proto.c]: might be good to log the error message here

File: nest/proto.c [hunk 4]
```diff
@@ -89,12 +97,58 @@ proto_log_state_change(struct proto *p)
...
     p->last_state_name_announced = NULL;
 }
 
+/**
+ * @proto: pointer to struct proto
+ * Function sends a UDP message to the BGP stat daemon notifying the state of the router
+ * This function will be called whenever the router state changes
+ */
+static void
+notify_bgp_stat_daemon(struct proto *p){
```
Comment (maxmzkr) [inline nest/proto.c]: please follow the spacing and hanging brackets style presented in the rest of the file

File: nest/proto.c [hunk 5]
```diff
@@ -57,8 +57,10 @@ static void proto_want_export_up(struct proto *p);
 static void proto_fell_down(struct proto *p);
 static char *proto_state_name(struct proto *p);
 static void notify_bgp_stat_daemon(struct proto *p);
+static int send_message_to_bgp_daemon(char *);
 
-#define BGP_PORT    4000
+
+#define BGP_PORT    4123
```
Comment (maxmzkr) [inline nest/proto.c]: bgp is really port 179. I'm thinking this shouldn't be called bgp_port

File: proto/bgp/bgp.c [hunk 6]
```diff
@@ -1310,7 +1310,7 @@ bgp_last_errmsg(struct bgp_proto *p)
     }
 }
 
-static const char *
+const char *
```
Comment (tacocat21) [inline proto/bgp/bgp.c:1313]: Removing static so I can use it on different c files

File: nest/proto.c [hunk 7]
```diff
@@ -83,18 +98,85 @@ static void
...
-      if (name != p->last_state_name_announced)
-	{
-	  p->last_state_name_announced = name;
-	  PD(p, "State changed to %s", proto_state_name(p));
-	}
-    }
-  else
+	    p->last_state_name_announced = name;
+	    PD(p, "State changed to %s", proto_state_name(p));
+  	  notify_bgp_stat_daemon(p);
```
Comment (maxmzkr) [inline nest/proto.c]: mix of tabs and spaces here

File: nest/proto.c [hunk 8]
```diff
@@ -83,18 +98,85 @@ static void
...
+    struct bgp_proto * p = (struct bgp_proto *)P;
+    const char * state = bgp_state_dsc(p);
+    char * name = P->name;
+    char message[1023 + 1];
+    struct bgp_config * cf = p->cf;
+    time_t time_of_change = time(NULL);
+    char buf[127 + 1];
+    struct tm * ts = gmtime(&time_of_change);
+    strftime(buf, sizeof(buf) - 1, "%Y-%m-%d-%H-%M-%S", ts);
+    bsnprintf(message, 1023, "%s$%I$%s$%s", name, cf->remote_ip, state, buf);
```
Comment (maxmzkr) [inline nest/proto.c]: should we do some sort of logging if our message doesn't fit inside the message?

Comment (maxmzkr) [inline nest/proto.c]: I still think that the message should include a type so that we can send multiple types of messages to our stats demon

Comment (zayfod) [reply nest/proto.c]: No. The buffers are supposed to be large enough.

Comment (zayfod) [reply nest/proto.c]: Agreed. I did not know there has been such discussion - changing...

Comment (maxmzkr) [reply nest/proto.c]: is there a guaranteed limit on name length?

Comment (zayfod) [reply nest/proto.c]: They are guaranteed to be "short". These are hardcoded protocol identifiers used in very few places in the code. The only names that are currently in use are `Kernel`, `Device`, `Static`, `RIP`, `RAdv`, `Pipe`, `OSPF`, `BGP`, `BFD`, and `Direct`.

There is not specific need to transfer the name actually...

Comment (zayfod) [reply nest/proto.c]: @maxmzkr, as discussed offline, I'll replace the current text protocol with a binary one on your request.

I'll document the solution in the design document first.

**Review by maxmzkr — APPROVED** [review]

**Pull request closed without merging**